File: typescript/src/Customer.ts
import {Address} from "@/Address";
import {ShoppingList} from "@/ShoppingList";
import {CustomerType} from "@/CustomerType";

export class Customer {
  constructor(public externalId: string | null,
              public masterExternalId: string | null,
              public address: Address | null = null,
              public preferredStore: string | null = null,
              public internalId: string | null = null,
              public name: string | null = null,
              public customerType: CustomerType | null = null,
              public companyNumber: string | null = null,
              public shoppingLists: ShoppingList[] = []) {
  }

  addShoppingList(consumerShoppingList: ShoppingList) {
    this.shoppingLists = [...this.shoppingLists, consumerShoppingList];
  }

  // Entity based equality
  isEqual(other: Customer) {
    return this.identityKey() === other.identityKey();
  }

  // May be used as Map keys
  identityKey() {
    return `${this.externalId}-${this.masterExternalId}-${this.companyNumber}`;
  }
}

File: java/src/main/java/codingdojo/model/Customer.java
package codingdojo.model;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@Accessors(chain = true)
public abstract class Customer {
    private String externalId;
    private String masterExternalId;
    private Address address;
    private String preferredStore;
    @Setter(AccessLevel.NONE)
    private List<ShoppingList> shoppingLists = new ArrayList<>();
    private String internalId;
    private String name;
    private CustomerType customerType;

    public Customer(CustomerType customerType) {
        this.customerType = customerType;
    }

    public void addShoppingList(ShoppingList consumerShoppingList) {
        List<ShoppingList> newList = new ArrayList<>(this.shoppingLists);
        newList.add(consumerShoppingList);
        this.shoppingLists = newList;
    }

}
